File: python/19-rapid-control-prototyping/main.py
#!/usr/bin/env python
# coding: utf-8

# # Ch19 快速控制原型(Rapid Control Prototyping, RCP)
# 
# **對應原書**:Chapter 19(原書用商用 RCP 硬體;這裡用 **ESP32 + Python** 全開源重現)
# **軟體替代**:商用 RCP 平台 → ESP32(即時迴路)+ Python(規劃/分析/調機),韌體見 `../../hardware/esp32/`
# 
# ## RCP 工作流程
# 1. **記錄**:ESP32 以 100 Hz 執行迴路,經序列埠回傳 CSV 遙測。
# 2. **辨識**:對開迴路步階記錄擬合一階模型 $K/(\tau s+1)$。
# 3. **設計**:λ-tuning 算 PI 增益。
# 4. **驗證**:先在 Python 模擬確認,再下載到 ESP32 實測。
# 5. **等價性**:Python 的 `DiscretePID` 與韌體 C 程式碼**逐樣本比對**(host 端編譯測試見 hardware/esp32/test_host)。
# 
# ## 遙測與命令協定(對應韌體)
# - 遙測(115200 baud):`millis,target,actual,pwm`
# - 命令:`SET,KP,2.0` / `SET,KI,1.0` / `SET,KD,0.05` / `SET,TARGET,500` / `STEP,100`(開迴路 PWM 步階)

# In[1]:


import sys, pathlib
sys.path.append(str(pathlib.Path(__file__).resolve().parent.parent))  # 讓腳本可從任意目錄執行
import numpy as np
import matplotlib.pyplot as plt
import control as ct
from common.control_helpers import (first_order, second_order, integrator, inertia,
                                    pi, pid, closed_loop, margins, step_metrics, bandwidth_hz)
from common.sim import (DiscretePID, MotorPlant, DCMotorPlant, TwoMassPlant,
                        Backlash, EncoderModel, Delay, saturate, quantize, coulomb)
from common.dsa import chirp_excitation, prbs_excitation, measure_frf, frf_of_system
from common.plots import bode_compare
np.set_printoptions(precision=4, suppress=True)


# In[2]:


import io
import pandas as pd
from scipy.optimize import curve_fit

# --- 步驟 1:模擬一段「ESP32 開迴路步階」遙測(實機上這段來自序列埠)---
K_true, tau_true = 3.0, 0.25        # 真實馬達:3 rpm / pwm-unit,τ=250 ms
fs_rcp = 100.0                       # ESP32 迴路 100 Hz
dt_rcp = 1 / fs_rcp
n = 300                              # 3 秒
rng = np.random.default_rng(1)
pwm_step = 100.0
t_log = np.arange(n) * dt_rcp
rpm = K_true * pwm_step * (1 - np.exp(-t_log / tau_true)) + rng.normal(0, 3.0, n)

lines = ['millis,target,actual,pwm']
for k in range(n):
    lines.append(f'{int(t_log[k]*1000)},0.0,{rpm[k]:.3f},{pwm_step:.1f}')
telemetry = '\n'.join(lines)
print('遙測前 5 行:')
print('\n'.join(telemetry.splitlines()[:5]))

# --- 步驟 2:解析 + 擬合 ---
df = pd.read_csv(io.StringIO(telemetry))
df['t'] = df['millis'] / 1000.0

def first_order_step(t, K, tau):
    return K * pwm_step * (1 - np.exp(-t / tau))

(K_fit, tau_fit), _ = curve_fit(first_order_step, df['t'], df['actual'], p0=[1.0, 0.1])
print(f'\n擬合:K={K_fit:.3f}(真值 {K_true}),tau={tau_fit*1000:.1f} ms(真值 {tau_true*1000:.0f} ms)')

plt.figure(figsize=(9, 4))
plt.plot(df['t'], df['actual'], '.', ms=2, alpha=0.5, label='telemetry')
plt.plot(df['t'], first_order_step(df['t'], K_fit, tau_fit), 'r-', label='fitted model')
plt.xlabel('Time [s]'); plt.ylabel('Speed [rpm]'); plt.legend(); plt.grid(alpha=0.3)
plt.title('Open-loop PWM step: system identification')
plt.show()


# ## 步驟 3–4:λ-tuning 設計 PI,先模擬驗證
# 
# 一階受控體的 λ-tuning(IMC):$k_p = \dfrac{\tau}{K\lambda}$、$k_i = k_p/\tau$,
# λ = 期望閉迴路時間常數(取 100 ms,約為 τ/2.5)。

# In[3]:


lam = 0.10
kp_rcp = tau_fit / (K_fit * lam)
ki_rcp = kp_rcp / tau_fit
print(f'λ-tuning:kp={kp_rcp:.4f} pwm/rpm,ki={ki_rcp:.4f}')

# 用「真實」受控體(擬合時不知道的)做離散閉迴路驗證,100 Hz、PWM 飽和 ±255
n2 = int(2.0 / dt_rcp)
t2 = np.arange(n2) * dt_rcp
target = 400.0
ctrl = DiscretePID(kp_rcp, ki_rcp, dt=dt_rcp, out_min=-255, out_max=255)
y_rpm = np.empty(n2)
y_state = 0.0
for k in range(n2):
    y_rpm[k] = y_state
    u = ctrl.step(target - y_state)
    y_state += (K_true * u - y_state) / tau_true * dt_rcp   # 真實一階馬達

mets = step_metrics(t2, y_rpm, target=target)
plt.figure(figsize=(9, 4))
plt.plot(t2, y_rpm); plt.axhline(target, color='k', ls='--', lw=0.8)
plt.axhline(target * (1 - np.exp(-1)), color='gray', ls=':', lw=0.8)
plt.xlabel('Time [s]'); plt.ylabel('Speed [rpm]'); plt.grid(alpha=0.3)
plt.title(f'Closed-loop check: OS={mets["overshoot_pct"]:.1f}%, λ design target ~{lam*1000:.0f} ms')
plt.show()
print('閉迴路指標:', {k: round(v, 4) for k, v in mets.items()})


# ## 步驟 5:韌體等價性驗證
# 
# 韌體 PID(見 `hardware/esp32/src/pid.h`)的差分方程與 `DiscretePID` 有一處已知差異:
# 韌體**先積分再算輸出**,`DiscretePID` 先算輸出再積分 —— 差一拍積分量 $k_i e T_s$。
# 下面逐樣本比對,確認差異就只有這一項(host 端 C++ 編譯比對見 `hardware/esp32/test_host/`)。

# In[4]:


class CStylePID:
    """逐行翻譯韌體 pid.h 的 pid_step():integ 先更新,導數不濾波。"""
    def __init__(self, kp, ki, kd, dt, out_min, out_max):
        self.kp, self.ki, self.kd, self.dt = kp, ki, kd, dt
        self.out_min, self.out_max = out_min, out_max
        self.integ, self.prev_e, self.first = 0.0, 0.0, True
    def step(self, e):
        self.integ += e * self.dt
        d = 0.0 if self.first else (e - self.prev_e) / self.dt
        self.first = False
        self.prev_e = e
        u = self.kp * e + self.ki * self.integ + self.kd * d
        return min(max(u, self.out_min), self.out_max)

rng = np.random.default_rng(3)
e_seq = rng.normal(0, 50, 500)                    # 隨機誤差序列 [rpm]
c_fw = CStylePID(kp_rcp, ki_rcp, 0.0, dt_rcp, -255, 255)
c_py = DiscretePID(kp_rcp, ki_rcp, 0.0, dt=dt_rcp, out_min=-255, out_max=255,
                   anti_windup=False)
u_fw = np.array([c_fw.step(e) for e in e_seq])
u_py = np.array([c_py.step(e) for e in e_seq])
diff = np.abs(u_fw - u_py)
bound = ki_rcp * np.abs(e_seq) * dt_rcp           # 理論差異上限:一拍積分量
print(f'最大差異 {diff.max():.4f} pwm-unit;理論上限 max(ki·|e|·Ts) = {bound.max():.4f}')
print(f'超出理論上限的樣本數:{(diff > bound + 1e-9).sum()} / {len(e_seq)}')


# In[5]:


# ✅ 驗證
assert abs(K_fit - K_true) / K_true < 0.05          # 辨識 K 誤差 < 5%
assert abs(tau_fit - tau_true) / tau_true < 0.10    # 辨識 tau 誤差 < 10%
assert mets['overshoot_pct'] < 15                   # λ-tuning 閉迴路溫和
assert abs(mets['steady_state_error']) < 5          # 穩態到位(rpm)
i63 = np.nonzero(y_rpm >= target * 0.632)[0][0]
assert abs(t2[i63] - lam) / lam < 0.5               # 63% 時間 ≈ λ
assert np.all(diff <= bound + 1e-9)                 # 韌體/Python 差異全在理論上限內
print('Ch19 驗證通過 ✅')


# ## 上實機(硬體到位後)
# 
# ```bash
# cd ../../hardware/esp32
# pio run -t upload && pio device monitor        # 燒錄 + 看遙測
# ```
# 1. `STEP,100` 記錄開迴路步階 → 用本 notebook 步驟 2 擬合你的馬達。
# 2. 把算出的 kp/ki 用 `SET,KP,…`/`SET,KI,…` 下發,`SET,TARGET,400` 看階躍。
# 3. 用第 2 章的 DSA:把 chirp 疊加在 target 上,量實機閉迴路 FRF。
# 
# ## 練習
# 1. 把 λ 掃 0.05/0.1/0.2,模擬三條 step response —— λ 就是「你想要多快」的旋鈕。
# 2. 對擬合殘差畫直方圖,雜訊是白的嗎?如果不是,漏掉了什麼動態?
# 3. 在模擬加入 20 ms 的傳輸延遲(序列埠+排程),λ=0.05 還穩嗎?對照第 4 章。
# 4. 讀 `hardware/esp32/src/pid.h`,找出它與 `DiscretePID` 的第二個差異(提示:導數濾波)。

File: cpp/19-rapid-control-prototyping/main.cpp
// Ch19 RCP:遙測 → 系統辨識 → λ-tuning → 離散閉迴路驗證 → 韌體 PID 等價性
//
// 與 Python 版不同之處:這裡直接 #include 韌體的 hardware/esp32/src/pid.h,
// 讓「會燒進 ESP32 的那份 C 程式碼」本身參與驗證。
#include <fstream>
#include <sstream>

#include "../../hardware/esp32/src/pid.h"
#include "../common/csd.hpp"
using namespace csd;

// 逐行翻譯 pid.h 的 pid_step()(double 版):integ 先更新,導數不濾波
struct CStylePID {
    double kp, ki, kd, dt, out_min, out_max, integ = 0, prev_e = 0;
    bool first = true;
    double step(double e) {
        integ += e * dt;
        double d = first ? 0.0 : (e - prev_e) / dt;
        first = false;
        prev_e = e;
        return saturate(kp * e + ki * integ + kd * d, out_min, out_max);
    }
};

// 兩參數 Levenberg–Marquardt:y ≈ K*pwm*(1 - exp(-t/tau))
std::pair<double, double> fit_first_order(const Vec& t, const Vec& y, double pwm, double K, double tau) {
    auto sse = [&](double k, double ta) {
        double s = 0;
        for (size_t i = 0; i < t.size(); ++i) {
            double r = y[i] - k * pwm * (1 - std::exp(-t[i] / ta));
            s += r * r;
        }
        return s;
    };
    double lam = 1e-3, cost = sse(K, tau);
    for (int it = 0; it < 200; ++it) {
        double a11 = 0, a12 = 0, a22 = 0, g1 = 0, g2 = 0;
        for (size_t i = 0; i < t.size(); ++i) {
            double e = std::exp(-t[i] / tau);
            double r = y[i] - K * pwm * (1 - e);
            double jK = pwm * (1 - e), jT = -K * pwm * e * t[i] / (tau * tau);
            a11 += jK * jK; a12 += jK * jT; a22 += jT * jT;
            g1 += jK * r; g2 += jT * r;
        }
        bool improved = false;
        while (lam < 1e12) {
            double b11 = a11 * (1 + lam), b22 = a22 * (1 + lam), det = b11 * b22 - a12 * a12;
            double dK = (b22 * g1 - a12 * g2) / det, dT = (b11 * g2 - a12 * g1) / det;
            if (tau + dT > 0) {
                double c = sse(K + dK, tau + dT);
                if (c < cost) {
                    bool done = std::fabs(dK) < 1e-12 * std::fabs(K) + 1e-15 && std::fabs(dT) < 1e-12 * tau + 1e-15;
                    K += dK; tau += dT; cost = c; lam *= 0.3; improved = true;
                    if (done) return {K, tau};
                    break;
                }
            }
            lam *= 10;
        }
        if (!improved) break;
    }
    return {K, tau};
}

int main() {
    // --- 步驟 1:模擬一段「ESP32 開迴路步階」遙測(實機上這段來自序列埠)---
    const double K_true = 3.0, tau_true = 0.25;  // 真實馬達:3 rpm / pwm-unit,τ=250 ms
    const double fs_rcp = 100.0, dt_rcp = 1 / fs_rcp;
    const int n = 300;
    const double pwm_step = 100.0;
    Rng rng(1);
    Vec noise = rng.normal(0, 3.0, n);
    const std::string tele_path = out_path("ch19_telemetry.csv");
    {
        std::ofstream f(tele_path);
        f << "millis,target,actual,pwm\n";
        for (int k = 0; k < n; ++k) {
            double tk = k * dt_rcp;
            char line[96];
            std::snprintf(line, sizeof line, "%d,0.0,%.3f,%.1f\n", int(tk * 1000),
                          K_true * pwm_step * (1 - std::exp(-tk / tau_true)) + noise[k], pwm_step);
            f << line;
        }
    }
    std::printf("遙測寫入 %s(格式同韌體序列輸出)\n", tele_path.c_str());

    // --- 步驟 2:解析 + 擬合 ---
    Vec t_log, actual;
    {
        std::ifstream f(tele_path);
        std::string line;
        std::getline(f, line);  // header
        while (std::getline(f, line)) {
            std::stringstream ss(line);
            std::string ms, tgt, act;
            std::getline(ss, ms, ',');
            std::getline(ss, tgt, ',');
            std::getline(ss, act, ',');
            t_log.push_back(std::stod(ms) / 1000.0);
            actual.push_back(std::stod(act));
        }
    }
    auto [K_fit, tau_fit] = fit_first_order(t_log, actual, pwm_step, 1.0, 0.1);
    std::printf("擬合:K=%.3f(真值 %.1f),tau=%.1f ms(真值 %.0f ms)\n", K_fit, K_true, tau_fit * 1000, tau_true * 1000);
    Vec fitted;
    for (double tk : t_log) fitted.push_back(K_fit * pwm_step * (1 - std::exp(-tk / tau_fit)));
    write_csv("ch19_identification.csv", {"t", "telemetry", "fitted"}, {t_log, actual, fitted});

    // --- 步驟 3:λ-tuning ---
    const double lam = 0.10;
    const double kp_rcp = tau_fit / (K_fit * lam), ki_rcp = kp_rcp / tau_fit;
    std::printf("λ-tuning:kp=%.4f pwm/rpm,ki=%.4f\n", kp_rcp, ki_rcp);

    // 用「真實」受控體(擬合時不知道的)做離散閉迴路驗證,100 Hz、PWM 飽和 ±255
    const int n2 = int(2.0 / dt_rcp);
    const double target = 400.0;
    Vec t2 = sample_times(n2, dt_rcp), y_rpm(n2);
    DiscretePID ctrl(kp_rcp, ki_rcp, 0.0, dt_rcp, -255, 255);
    double y_state = 0.0;
    for (int k = 0; k < n2; ++k) {
        y_rpm[k] = y_state;
        double u = ctrl.step(target - y_state);
        y_state += (K_true * u - y_state) / tau_true * dt_rcp;  // 真實一階馬達
    }
    StepMetrics mets = step_metrics(t2, y_rpm, target);
    std::printf("閉迴路指標:rise=%.4f OS=%.4f%% settle=%.4f sse=%.4f\n", mets.rise_time, mets.overshoot_pct,
                mets.settling_time, mets.steady_state_error);
    write_csv("ch19_closed_loop.csv", {"t", "rpm"}, {t2, y_rpm});

    // --- 步驟 4a:韌體演算法 vs 教材 DiscretePID(差異應全在一拍積分量內)---
    Rng rng2(3);
    Vec e_seq = rng2.normal(0, 50, 500);  // 隨機誤差序列 [rpm]
    CStylePID c_fw{kp_rcp, ki_rcp, 0.0, dt_rcp, -255, 255};
    DiscretePID c_py(kp_rcp, ki_rcp, 0.0, dt_rcp, -255, 255, 0.0, false);
    Vec u_fw, diff, bound;
    int over = 0;
    for (double e : e_seq) {
        u_fw.push_back(c_fw.step(e));
        diff.push_back(std::fabs(u_fw.back() - c_py.step(e)));
        bound.push_back(ki_rcp * std::fabs(e) * dt_rcp);  // 理論差異上限:一拍積分量
        over += diff.back() > bound.back() + 1e-9;
    }
    std::printf("最大差異 %.4f pwm-unit;理論上限 max(ki·|e|·Ts) = %.4f\n", vmax(diff), vmax(bound));
    std::printf("超出理論上限的樣本數:%d / %zu\n", over, e_seq.size());

    // --- 步驟 4b:真正的韌體 pid.h(float)vs double 翻譯版 ---
    pid_t_ fw;
    pid_init(&fw, float(kp_rcp), float(ki_rcp), 0.0f, float(dt_rcp), -255.0f, 255.0f);
    double fw_worst = 0.0;
    for (size_t k = 0; k < e_seq.size(); ++k)
        fw_worst = std::max(fw_worst, std::fabs(double(pid_step(&fw, float(e_seq[k]))) - u_fw[k]));
    std::printf("韌體 pid.h(float32)vs double 版最大差異 %.2e pwm-unit\n", fw_worst);

    Checker chk("Ch19");
    CHECK(chk, std::fabs(K_fit - K_true) / K_true < 0.05);         // 辨識 K 誤差 < 5%
    CHECK(chk, std::fabs(tau_fit - tau_true) / tau_true < 0.10);   // 辨識 tau 誤差 < 10%
    CHECK(chk, mets.overshoot_pct < 15);                           // λ-tuning 閉迴路溫和
    CHECK(chk, std::fabs(mets.steady_state_error) < 5);            // 穩態到位(rpm)
    long i63 = first_index(y_rpm.size(), [&](size_t i) { return y_rpm[i] >= target * 0.632; });
    CHECK(chk, std::fabs(t2[i63] - lam) / lam < 0.5);              // 63% 時間 ≈ λ
    CHECK(chk, over == 0);                                         // 韌體/教材差異全在理論上限內
    CHECK(chk, fw_worst < 1e-3);                                   // float32 韌體與 double 版一致
    return chk.finish();
}
